ThirdPartyAdapters/vungle/vungle/src/main/java/com/vungle/mediation/VungleBannerAdapter.java

```diff
@@ -76,22 +110,19 @@ void setAdLayout(@NonNull RelativeLayout adLayout) {
     this.mAdLayout = new WeakReference<>(adLayout);
   }
 
-  void setVungleListener(@Nullable VungleListener vungleListener) {
-    this.mVungleListener = new WeakReference<>(vungleListener);
-  }
-
-  @Nullable
-  private VungleListener getVungleListener() {
-    return mVungleListener.get();
-  }
-
   boolean isActive() {
     return mAdLayout.get() != null;
   }
 
-  void requestBannerAd(@NonNull Context context, @NonNull String appId) {
+  void requestBannerAd(@NonNull Context context, @NonNull String appId,
+      @NonNull MediationBannerListener mediationBannerListener,
+      @NonNull MediationBannerAdapter mediationBannerAdapter) {
     Log.d(TAG, "requestBannerAd: " + this);
     mPendingRequestBanner = true;
+
+    bannerAdapter = mediationBannerAdapter;
+    bannerListener = mediationBannerListener;
```

Comment:
```suggestion
    bannerListenerWeak = new WeakReference<>(mediationBannerListener);
```

Reply:
AdMob rollbacked this PR https://github.com/googleads/googleads-mobile-android-mediation/pull/271 due to banner click/open/close callbacks not being forwarded. It seems weak mediationBannerListener reference easy GCed and causes that issue. So my suggestion is to keep mVungleBannerListener implementation in VungleInterstitialAdapter and mVungleListener in VungleBannerAdapter.

Reply:
@HuFangshuai agreed. Let's suggest for them not to make the changes atm since based on how we cleanup our banner ads we'll leak the listeners atm, though if they wanted to go forward w/ the change, we'll suggest @alexander-gubov commit suggestions.

I did some digging and it looks like in general design pattern wise google might prefer no hard references to mediationbannerlistener in adapterclasses:
https://github.com/googleads/googleads-mobile-android-mediation/blob/master/ThirdPartyAdapters/applovin/applovin/src/main/java/com/applovin/mediation/ApplovinAdapter.java

https://github.com/googleads/googleads-mobile-android-mediation/blob/master/ThirdPartyAdapters/facebook/facebook/src/main/java/com/google/ads/mediation/facebook/FacebookAdapter.java

It would help to ask them if there were any design patterns that google preferred so that we didn't have to handle these issues in suddens PRs.

Reply:
@HuFangshuai @jiancchen @alexander-gubov I'm from Google and would be happy to talk a bit more about these design patterns.

In a lot of the PRs we've reviewed from the Vungle team, and ones you review from us, we do seem to be at odds about how closely we need to track weak references to avoid memory leaks. It's important to note the differences between iOS and Android paradigms - I think the Vungle team might be trying to apply iOS paradigms to Android. In the iOS world, it is common best practice to make delegate variables `weak` to avoid retain cycles. Java garbage collection, however, [is able to clean up 2 object that have a retain cycle if no other objects reference them](https://stackoverflow.com/questions/1910194/how-does-java-garbage-collection-work-with-circular-references#:~:text=yes%20Java%20Garbage%20collector%20handles,them%20at%20its%20own%20root.), so it's not that important to make listeners weak.

As such, it's common practice in Android for callers to not keep references to their own listeners. Consider this simple example from the Android [Button](https://developer.android.com/reference/android/widget/Button) Javadocs:

```
final Button button = findViewById(R.id.button_id);
button.setOnClickListener(new View.OnClickListener() {
    public void onClick(View v) {
        // Code here executes on main thread after user presses button
    }
});
```

The caller passes in anonymous class. If the button class doesn't keep a strong reference to that View.OnClickListener, it will never get called. So unlike iOS, it is expected that the library accepting the listener keep a strong reference. Applying that to the mediation API, the `MediationBannerListener` object provided to Vungle at request time is not kept by the Google Mobile Ads SDK. It's expected that the adapter holds onto it. The reason the AppLovin adapter doesn't keep a direct reference is because it's wrapped in an object that's given to AppLovin's SDK, which is keeping a strong reference to it. The Facebook adapter linked above [does keep a strong reference to the listener](https://github.com/googleads/googleads-mobile-android-mediation/blob/master/ThirdPartyAdapters/facebook/facebook/src/main/java/com/google/ads/mediation/facebook/FacebookAdapter.java#L81-L82).

For keeping a reference to the Adapter object itself, for example VungleInterstitialAdapter, that is also generally fine since Java can garbage collect cyclic references. If the Google Mobile Ads SDK sets VungleInterstitialAdapter to null, it will garbage collect VungleBannerAdapter and other references which have no other code holding it. I suppose weak reference on the adapter can also work, as the Google Mobile Ads SDK will hold onto the adapter until it no longer needs it, but I think it just makes the code more complex to have to pack and unpack a weak reference every time.

Where we DO keep weak references is for objects that have a lifecycle that's longer than the lifecycle of an adapter request. A context (particularly if it's an activity context) is one example. The other is when there are `static` references to adapters, as a static reference will prevent garbage collection from running when the Google Mobile Ads SDK drops the reference. We do this for [rewarded ads](https://github.com/googleads/googleads-mobile-android-mediation/blob/master/ThirdPartyAdapters/vungle/vungle/src/main/java/com/google/ads/mediation/vungle/VungleMediationAdapter.java#L46) on Vungle, where we need to map vungle's rewarded callbacks to the appropriate adapter and make sure that only 1 adapter per placement can respond to a request.

I do hope you'll reconsider keeping the listener as a strong reference, and work with us to approve this refactor so we can more cleanly separate banner code from interstitial code.

Reply:
@ericleich
The issue is that we keep VungleBannerAdapter in a separate map inside VungleManager to be able to call com.vungle.warren.VungleBanner#destroyAd in the absence of com.google.android.gms.ads.mediation.MediationAdapter#onDestroy call from AdMob side
If we remove this map of VungleBannerAdapter, both VungleBannerAdapter & VungleMediationAdapter can be garbage collected but since nobody calls "destroyAd()"  the corresponding PlacementId remains blocked inside Vungle logic

Reply:
@jiancchen changes are ok, we just need to avoid any strong reference to VungleMediationAdapter from VungleBanner (even via BannerListener) because it is how we detect missed onDestroy in VungleManager#cleanLeakedBannerAdapters

Reply:
So, this situation is created due to VungleManager keeping a strong HashMap to VungleBannerAdapter references ([code](https://github.com/googleads/googleads-mobile-android-mediation/blob/master/ThirdPartyAdapters/vungle/vungle/src/main/java/com/vungle/mediation/VungleManager.java#L26)). Does it have to keep a strong reference to adapter objects?  How about something that isn't tied to the adapter lifecycle? We could introduce a new data class with that just represents a banner with the request ID tied to it:

```
class VungleBannerAd {
  private String uniquePubRequestId;
  @Nullable private VungleBanner mVungleBannerAd;
  @Nullable private VungleNativeAd mVungleNativeAd;

  public VungleBanner(@NonNull VungleBanner bannerAd, @NonNull String uniquePubRequestId);
  public VungleBanner(@NonNull VungleNativeAd nativeAd, @NonNull String uniquePubRequestId);
  public String getUniqueRequestId();
  @Nullable public String getVungleBanner();
  @Nullable public String getVungleNativeAd();
}
```

And we could then keep a HashMap<String, VungleBannerAd> instead? No more ties to adapters or Context objects and so no more risk of memory leaks.

Reply:
@ericleich Yes, we don't need a strong reference to adapter from the map, we only need an ability to call destroy API on mVungleBannerAd/mVungleNativeAd. This code can be separated.
The important part is to keep attach/detach logic to remove the reference from mVungleBannerAd/mVungleNativeAd to the adapter via View parent references. Those functions can be moved to 'VungleBannerAd'

Reply:
@alexander-gubov I'd like to propose that we frst do a PR then to see if we can make the VungleManager HashMap a <String, VungleBannerAd> type, and if we can align on those changes, this banner refactor PR should fall back into place. I'll circle back with Joshua on this and propose to him we send you a PR for that first.

Reply:
HI @alexander-gubov , @jiancchen, @HuFangshuai 
We have created https://github.com/googleads/googleads-mobile-android-mediation/pull/304 that refactored the VungleManager class as a result of this discussion.
Can you take a look?